eventbridge/eventbridge.go

```diff
@@ -0,0 +1,173 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package eventbridge
+
+import (
+	"context"
+
+	// We need the typeurl register calls that occur in the init of this package in order to be able to unmarshal events
+	// in the Republish func below
+	_ "github.com/containerd/containerd/api/events"
+
+	// Even though we are following the v2 runtime model, we are currently re-using a struct definition (Envelope) from
+	// the v1 event API
+	eventapi "github.com/containerd/containerd/api/services/events/v1"
+
+	"github.com/containerd/containerd/events"
+	"github.com/containerd/ttrpc"
+	"github.com/containerd/typeurl"
+	"github.com/gogo/protobuf/types"
+)
+
+const (
+	getterServiceName  = "aws.firecracker.containerd.eventbridge.getter"
+	getEventMethodName = "GetEvent"
+)
+
+// The Getter interface provides an API for retrieving containerd events from a remote event source, such as an
+// exchange. It exists separately from containerd's existing EventClient and EventServer interfaces in order to support
+// retrieving events via long-polling as opposed to a streaming model. This allows us to use TTRPC for the
+// implementation, which currently does not support streaming.
+type Getter interface {
+	GetEvent(ctx context.Context) (*eventapi.Envelope, error)
+}
+
+type getterService struct {
+	eventChan <-chan *events.Envelope
+	errChan   <-chan error
+}
+
+// NewGetterService returns a server-side implementation of the Getter interface. Given an existing event source, it
+// will subscribe to that source with no filters. Each GetEvent call pops and returns an event buffered in the
+// subscription's channel of published events. If no event is buffered, it blocks until one is.
+func NewGetterService(ctx context.Context, eventSource events.Subscriber) Getter {
+	eventChan, errChan := eventSource.Subscribe(ctx)
+	return &getterService{
+		eventChan: eventChan,
+		errChan:   errChan,
+	}
+}
+
+// GetEvent pops and returns an event buffered in the service's subscription channel.
+func (s *getterService) GetEvent(ctx context.Context) (*eventapi.Envelope, error) {
+	select {
+	case receivedEnvelope := <-s.eventChan:
+		return &eventapi.Envelope{
+			Timestamp: receivedEnvelope.Timestamp,
+			Namespace: receivedEnvelope.Namespace,
+			Topic:     receivedEnvelope.Topic,
+			Event:     receivedEnvelope.Event,
+		}, nil
+	case err := <-s.errChan:
+		return nil, err
+	case <-ctx.Done():
+		if ctxErr := ctx.Err(); ctxErr != context.Canceled {
+			return nil, ctxErr
+		}
+		return nil, nil
```

Comment:
I would rather return an error and make client responsible whether he wants to ignore `context.Canceled`.
When `GetEvent` returns `nil` error I expect that I can use envelope safely. Or at least this should be mentioned in godoc.

Reply:
Yeah I see your point, I went back and forth on this a bit. My original reasoning was that because the caller is providing context, they can be aware if it's canceled just as easily as the callee and thus shouldn't be forced to always check `if err != context.Canceled`. Err should just indicate that something went wrong with the callee's logic specifically.

However, you make a good a point that its pretty standard to assume that if err is nil you can go ahead and use the primary return value(s). Given that's a much more common assumption and doesn't change anything functionally, I'll switch to what you're suggesting.